sql/src/main/java/io/crate/jobs/ExecutionSubContext.java

```diff
@@ -25,9 +25,32 @@
 
 public interface ExecutionSubContext {
 
+    /**
+     * Add a callback to listen to events and failures of this subcontext. This callback is required to be set before
+     * any of the other methods are called.
+     *
+     * @param contextCallback
+     */
     void addCallback(ContextCallback contextCallback);
+
+    /**
+     * In the prepare phase implementations of this interface can allocate any resources.
+     *
+     * In this phase failures must not be propagated to downstream phases directly, but instead being reported by
+     * calling {@link io.crate.jobs.ContextCallback#onClose(Throwable, long)} with the throwable.
+     */
+    void prepare();
+
+    /**
+     * In the start phase implementations of this interface are required to start any executors.
+     *
+     * In this phase failures must not be propagated to downstream phases directly, but instead being reported by
```

Comment:
Maybe clarify that the started executors are allowed to propagate failures to the downstream